engine/execution/computation/computer/computer.go

```diff
@@ -306,6 +306,26 @@ func numberOfTransactionsInBlock(collections []*entity.CompleteCollection) int {
 	return numTxns
 }
 
+// selectChunkConstructorForProtocolVersion selects a [flow.Chunk] constructor to
+// use when constructing the [flow.ExecutionResult] for the input block. We select
+// based on the protocol version at the input block. When we process the version upgrade
+// event to protocol version 2, we begin populating the new [flow.ChunkBody.ServiceEventCount]
+// field.
+// Deprecated:
+// TODO(mainnet27, #6773): remove this function https://github.com/onflow/flow-go/issues/6773
+func (e *blockComputer) selectChunkConstructorForProtocolVersion(blockID flow.Identifier) (flow.ChunkConstructor, error) {
+	ps, err := e.protocolState.AtBlockID(blockID).ProtocolState()
+	if err != nil {
+		return nil, err
+	}
+	version := ps.GetProtocolStateVersion()
+	if version < 2 {
+		return flow.NewChunk_ProtocolVersion1, nil
+	} else {
+		return flow.NewChunk, nil
+	}
```

Comment:
would prefer to make this as strict as possible. This code might be living for multiple months until we spork into mainnet 27. There might be more updates in between now and the spork. I would like to avoid making assumptions about the nature of future updates (e.g. that they do not change the chunk format). 

```suggestion
	switch version {
	case 1:
		return flow.NewChunk_ProtocolVersion1, nil
	case 2:
		return flow.NewChunk, nil
	default:
		return nil, fmt.Errorf("unsupported chunk version: %d", version)
	}
```

Reply:
> There might be more updates in between now and the spork. I would like to avoid making assumptions about the nature of future updates

Every other part of our codebase already implicitly makes assumptions about the nature of future updates, by virtue of not having every component, model, etc. tied to a protocol version number. It will always ultimately be the responsibility of the upgrade implementor to make sure it is safe and compatible with the existing logic and data structures.

I don't have a problem with the suggestion here on its own, but I don't think it represents a sustainable strategy for managing version-linked code changes. Effectively all of the time, when we fix a bug or add a feature, we intend for that logical change to persist in perpetuity, until it is actively changed again. I expect the same to be true for bug fixes and features introduced through a protocol version upgrade. 

If in the future an engineer is changing the Chunk data structure for a protocol upgrade, they already need to carefully investigate and update logic where the Chunk is being used. This error will help them find one of the codepaths they need to change, but there are many more. Had they failed to notice this codepath without the error, they very likely would also have failed to notice other codepaths that don't have a version switch statement. **Conclusion:** sometimes helpful, but far from sufficient as a safety measure. 

By contrast, if a future protocol upgrade changes something completely unrelated to the Chunk, they will need to update a growing number of these switch statements (minor inconvenience). Now suppose our version-based switch statement is in an infrequently accessed codepath (say a slashing challenge). The engineer tests their changes on the new protocol version, but their testing doesn't include our infrequently accessed codepath on the new version. Now we have introduced a severe bug that might only appear in production, after the version upgrade is complete. **Conclusion:** usually only a minor annoyance, but potential for severe bugs.